test/storage_accessible.js

```diff
@@ -1,51 +1,93 @@
-    await instance.setQuxKeyValue(42, 69)
-    assert.equal(69, await instance.getStorageAt(keccak([42, await instance.SLOT_QUX()]), 1))
+      const data = await instance.getStorageAt(keccak([slot]), length)
+      assert.equal(1, fromHex(data.slice(2, 66)))
+      assert.equal(2, fromHex(data.slice(66, 130)))
+    })
+    it('can read mappings', async () => {
+      const instance = await StorageAccessibleWrapper.new()
+      await instance.setQuxKeyValue(42, 69)
+      assert.equal(69, await instance.getStorageAt(keccak([42, await instance.SLOT_QUX()]), 1))
+    })
+
+    it('can read structs', async () => {
+      const instance = await StorageAccessibleWrapper.new()
+      await instance.setFoobar(19, 21)
+
+      const packed = await instance.getStorageAt(await instance.SLOT_FOOBAR(), 10)
+      assert.equal(19, fromHex(packed.slice(2, 66)))
+      assert.equal(21, fromHex(packed.slice(66, 130)))
+    })
   })
 
-  it('can read structs', async () => {
-    const instance = await StorageAccessibleWrapper.new()
-    await instance.setFoobar(19, 21)
+  describe('simulateDelegatecall', async () => {
+    it('can invoke a function in the context of a previously deployed contract', async () => {
+      const instance = await StorageAccessibleWrapper.new()
+      await instance.setFoo(42)
+
+      // Deploy and use reader contract to access foo
+      const reader = await ExternalStorageReader.new()
+      const getFooCall = reader.contract.methods.getFoo().encodeABI()
+      const result = await instance.simulateDelegatecall.call(reader.address, getFooCall)
+      assert.equal(42, fromHex(result))
+    })
+
+    it('can simulate a function with side effects (without executing)', async () => {
+      const instance = await StorageAccessibleWrapper.new()
+      await instance.setFoo(42)
+
+      // Deploy and use reader contract to simulate setting foo
+      const reader = await ExternalStorageReader.new()
+      const replaceFooCall = reader.contract.methods.setAndGetFoo(69).encodeABI()
+      let result = await instance.simulateDelegatecall.call(reader.address, replaceFooCall)
+      assert.equal(69, fromHex(result))
+
+      // Make sure foo is not actually changed
+      const getFooCall = reader.contract.methods.getFoo().encodeABI()
+      result = await instance.simulateDelegatecall.call(reader.address, getFooCall)
+      assert.equal(42, fromHex(result))
+    })
+
+    it('can simulate a function that reverts', async () => {
+      const instance = await StorageAccessibleWrapper.new()
 
-    const packed = await instance.getStorageAt(await instance.SLOT_FOOBAR(), 10)
-    assert.equal(19, fromHex(packed.slice(2, 66)))
-    assert.equal(21, fromHex(packed.slice(66, 130)))
+      const reader = await ExternalStorageReader.new()
+      const doRevertCall = reader.contract.methods.doRevert().encodeABI()
+      truffleAssert.reverts(instance.simulateDelegatecall.call(reader.address), doRevertCall)
```

Comment:
Shouldn't it be:
```suggestion
      truffleAssert.reverts(instance.simulateDelegatecall.call(reader.address, doRevertCall))
```
?

Reply:
Totally. Great catch. Still passes.

Reply:
This is one of those subtle issues that TypeScript would have helped find :stuck_out_tongue: